# 第六章 异常处理
* 1.体系结构
```java
   java.lang.Object
|---java.lang.Throwable
|---java.lang.Error：错误，java程序无能为力，不显式的处理
|---java.lang.Exception：异常，需要进行处理
|---RuntimeException:运行时异常
|---ArrayIndexOutOfBoundsException /NullPointerException /
     ArithmeticException  /ClassCastException
|---非RuntimeException:编译时异常
```
* 2.因为java程序分为javac.exe和java.exe两个过程，在每个过程中，都有可能出现异常。故分为编译时异常，运行时异常
    * 2.1对于运行时异常比较常见，可以不显式的来处理。
    * 2.2对于编译时异常，必须要显式的处理
编译时异常，不是说有异常才处理，而是存在异常的隐患，必须在编译时，提示程序，万一出现异常，如何处理！

* 3.如何处理异常？<br>
java中的“抓抛模型”<br>
    * 1."抛"：当我们执行代码时，一旦出现异常，就会在异常的代码处生成一个对应的异常类型的对象，并将此对象抛出。（自动抛出   / 手动抛出）
    * 一旦抛出此异常类的对象，那么程序就终止执行
    * 此异常类的对象抛给方法的调用者。
    * 2."抓"：抓住上一步抛出来的异常类的对象。如何抓？即为异常处理的方式
**Java 提供了两种方式用来处理一个异常类的对象。**
    * **处理的方式一：**
```java
try{
//可能出现异常的代码
}catch(Exception1 e1){
//处理的方式1
}catch(Exception2 e2){
//处理的方式2
}finally{
//一定要执行的代码
}
```
注：
 *  1.try内声明的变量，类似于局部变量，出了try{ }语句，就不能被调用
 * 	2.finally是可选的。
 * 	3.catch语句内部是对异常对象的处理。<br>
`getMessage();   printStackTrace();`
 * 	4.可以有多个catch语句，try中抛出的异常对象从上往下去匹配catch中的异常类的类型，
 * 	一旦满足就执行catch中的代码。执行完，就跳出其后的多条catch语句
 * 	5.如果异常处理了，那么其后的代码继续执行。
 * 	6.若catch中多个异常类型是"并列"关系，谁上谁下都可以。
 * 	若catch中多个异常类型是"包含"关系，须将子类放在父类的上面，进行处理。否则报错！
 * 	7.finally中存放的是一定会被执行的代码，不管try中、catch中是否仍有异常未被处理，以及是否有return语句。
 * 	8.try-catch是可以嵌套的。

    * **处理方式二：**<br>
    在方法的声明处，显式的使用throws +异常类型
```java
public void method1() throws Exception1 e1,Exception2 e2{
//可能出现异常（尤其是编译时异常，一定要处理）
}

public void method2() throws Exception e1,Exception e2{
method1();;
}

public void method3(){
try{
method2();
}catch(Exception1  e1){
System.out.println(e1.getMessage());
}catch(Exception2  e2){
System.out.println(e2.getMessage());
}
}

public static void main(String[] args){
对象1.method3();//不会再出现上述的Exception1和Exception2的异常！
}
```
* 4.如何手动的抛出一个异常？<br>
   在方法的内部，可以使用 throw  + 异常类对象 ，来手动的抛出一个异常！
```java
//比较两个圆的半径的大小。
	public int compareTo(Object obj) throws Exception{
		if(this==obj){
			return 0;
		}else if(obj instanceof Circle){
			Circle c=(Circle)obj;
			if(this.radius>c.radius){
				return 1;
			}else if(this.radius==c.radius){
				return 0;
			}else{
				return -1;
			}
		}else{
			//return -2;
			//手动的抛出一个异常
			//throw new RuntimeException("传入的类型有误!");
			//throw new Exception("传入的类型有误!");
			throw new MyException("传入的类型有误!");
		}
	}
```
* 5.如何自定义一个异常类？<br>
手动的抛出一个异常，除了抛出的是现成的异常类的对象之外，还可以抛出一个自定义的异常类的对象！<br>
如何自定义一个异常类呢？<br>
```java
/*
 * 1.自定义的异常类继承现有的异常类
 * 2.提供一个序列号，提供几个重载的构造器
 */
public class MyException extends Exception {
	static final long serialVersionUID = -70348975766939L;

	public MyException() {

	}

	public MyException(String msg) {
		super(msg);
	}
}
```
*  6. “5个关键字搞定异常处理！” 其中，要区分：throw与throws的区别？
